onnx/onnx_backend.h

```diff
@@ -0,0 +1,801 @@
+   * Possible values:
+   *     ONNX_MEMORY_TYPE_CPU
+   */
+  onnxEnum memoryType;
+  /**
+   * Number of dimensions in the tensor.
+   * Must be between 0 (for a scalar) and ONNX_TENSOR_DIMS_MAX.
+   */
+  uint32_t dimensions;
+  /**
+   * Dimensions of the tensor.
+   */
+  uint64_t shape[ONNX_TENSOR_DIMS_MAX];
+  /**
+   * Pointers to tensor data.
+   */
+  onnxPointer buffer;
+} onnxTensorDescriptor;
+
+/* Function pointer declarations for dynamic loading */
+typedef uint32_t(ONNX_ABI* onnxGetNumBackendsFunction)(void);
+typedef onnxStatus(ONNX_ABI* onnxGetBackendInfoFunction)(
+    uint32_t index,
+    void* backendInfo,
+    uint32_t backendInfoVersion);
+typedef onnxStatus(ONNX_ABI* onnxGetBackendCompatibilityFunction)(
+    uint32_t index,
+    size_t onnxModelSize,
+    const void* onnxModel);
+typedef onnxStatus(ONNX_ABI* onnxInitBackendFunction)(
+    uint32_t index,
+    const uint64_t* auxPropertiesList,
+    onnxBackend* backend);
+typedef onnxStatus(ONNX_ABI* onnxReleaseBackendFunction)(onnxBackend backend);
+typedef onnxStatus(ONNX_ABI* onnxInitGraphFunction)(
+    onnxBackend backend,
+    size_t onnxModelSize,
+    const void* onnxModel,
+    uint32_t weightsCount,
+    const onnxTensorDescriptor* weightDescriptors,
+    onnxGraph* graph);
+typedef onnxStatus(ONNX_ABI* onnxSetGraphIOFunction)(
+    onnxGraph graph,
+    uint32_t inputsCount,
+    const onnxTensorDescriptor* inputDescriptors,
+    uint32_t outputsCount,
+    const onnxTensorDescriptor* outputDescriptors);
+typedef onnxStatus(ONNX_ABI* onnxRunGraphFunction)(onnxGraph graph);
+typedef onnxStatus(ONNX_ABI* onnxReleaseGraphFunction)(onnxGraph graph);
+
+/**
+ * Query the number of available backends for ONNX graphs.
+ *
+ * ONNX backend is a combination of software layer and hardware device used to
+ * run an ONNX graph. The same software layer may expose multiple backends (e.g.
+ * one ONNX backend for each GPU in the system, or one ONNX backend for GPU and
+ * another for CPU, both implemented in the same software). Backends implemented
+ * in the same software, but targeting different devices (e.g. "MyNN" for CPU
+ * and "MyNN" for GPU) are counted separately.
+ */
+ONNX_PUBLIC uint32_t ONNX_ABI ONNX_SYMBOL_NAME(onnxGetNumBackends)(void);
+
+/**
+ * Query high-level information about the backend and its target device.
+ *
+ * ONNX backend is a combination of software layer and hardware device used to
+ * run an ONNX graph. The same software layer may expose multiple backends (e.g.
+ * one ONNX backend for each GPU in the system, or one ONNX backend for GPU and
+ * another for CPU, both implemented in the same software).
+ *
+ * @param index - index of the backend to query.
+ * @param[out] backendInfo - pointer to backend information structure, either
+ *                           onnxBackendCoreInfo (if backendInfoType is
+ *                           ONNX_BACKEND_INFO_TYPE_CORE), or
+ *                           onnxBackendPerformanceInfo (if backendInfoType is
```

Comment:
I'd rather put these into two different functions which are more strictly typed. One returns the `onnxBackendCoreInfo`, the other function returns the `onnxBackendPerformanceInfo`. This way, we also get rid of the `backendInfoType` flag and it's safer to user. Also, the parameter can then be more strictly typed instead of just `void*`.